README.md
# élelmiszer MAD
MAD utilizes distinct features from IBM watson using R and Python and display latest trends,statistics from social media accounts

### Few words from Andrew Zimmern a food traveler and critic. 
       
>I think Yelp is neither good nor bad for the food industry. I find it useless. 
>If you're aggregating a lack of expertise, then   when I plug in “best sushi bar in Los Angeles,” Yelp doesn't help me at all.
>So, if you are a huge food geek like me that really believes in quality — not expensive food, just quality.
>Above all other things — quality. If you're into quality, Yelp is not for you.

>Zimmern says that when he’s looking for something good to eat in a city he’s visiting, 
>he checks out the Twitter and Instagram feeds of local food writers and chefs to find out the places and dishes that 
>they’re obsessing over, and he builds his list from there. 
		
>The author/TV host also remarks: “Yelp’s not even good for looking up the restaurant’s phone number because you know on the site, 
>they just want you to read their reviews and look at their ads.

## Problem statement:
As it was mentioned, considering reviews from one source to get information regarding the local delicacies, is not much trustworthy
Most of us who walk in the shoes of Mr. Zimmern try to collect information from as many sources as possible.

## Proposed solution: 
Through this application, we are trying to utilize the data from three different sources YELP , TWITTER and INSTAGRAM analyze the data and gain insights.

## Approaching the problem: 

* Analyze text based on what the user has searched for in the textbox "What food are you in mood for today?".
The text is either fed as a Hashtag/term/user in the twitter. 
Metadata such as  entities, keywords, categories, relations and semantic roles, sentiment and emotion are extracted from the content as they provide insights on tweets posted in the Twitter. 
After verifying whether the posted term is in a tweet and belong to the category "Food",  which is one of the primary categories in IBM Watson's category classifier, we filter and approximate the location , keywords familiar or tag along with the searched term. 

* Similarly gathering the data from Yelp through the API available, gather reviews based on the restaurant. Since not all reviews are equal in favor of the dish , to weigh the true dish value ,and in loss of expertise reviews we consider many other factors and aggregat them. We search for past tweets/reviews from twitter and yelp respectively posted by food critics. Their reviews are assigned a major weight compared to the other reviews.

* Using the endpoints provided by the Watson API , we can gain personality insights and analyze text for yelp data too.

## Conclusion
The end result would be calculating , if the restaurant or location has been mentioned in any tweet , tweet's emotion and sentiment
if any dish mentioned in yelp in a review , it's emotion , sentiment 

## Limitations to the application







history.txt
fluidRow(column(3, verbatimTextOutput("value"))),
tabsetPanel(id = 'dataset',
tabPanel("table_summary", DT::dataTableOutput("table_summary")),
tabPanel("df",DT::dataTableOutput("table_tweets"))
)
)
)
)
# Define server logic required to draw a histogram
server <- function(input, output) {
twitt_df <- twitter_stuff(1000)
twFeed_csv<- write.csv(twitt_df,file = "savedTweets.csv",append = TRUE)
df = twitt_df[sample(nrow(twitt_df),500),]
bind_Signal <- data.frame() #take tweet id from tweet dataframe and bind POST content with the tweet_id
bind_Signal_to_df<-data.frame()
#choose columns to display
jsonSignal<-give_me_a_json_damnit(df, bind_Signal, bind_Signal_to_df)
output$table_summary<- DT::renderDataTable({DT::datatable(jsonSignal)})
output$table_tweets<- DT::renderDataTable({DT::datatable(df,options = list(lengthMenu = c(50, 300, 500)))})
output$value <- renderPrint("Done")
}
# Run the application
shinyApp(ui = ui, server = server)
#add twitter keyword as input parameter from shiny textbox at later stage
library(twitteR)
library(plyr)
twitter_stuff<- function(numTweets){
#Setup the Twitter Account and pass the tokens
setup_twitter_oauth(APIKey,APISecret,accessToken,accessTokenSecret)
#input text "Dosa OR Dhosa OR #Dosa OR #Dhosa
tweets_list <- twitteR::getUser(user = "@Deals4Every")
#getUser<- twitteR::getUser("@Deals4Every")
tweets_text = laply(tweets_list, function(t) t$getText())
tweets_clean <- clean.text(tweets_text)
#convert the tweets_text object to a data frame
#make data frame
df <- do.call("rbind", lapply(tweets_list, as.data.frame))
df["text"] = tweets_clean
return(df)
}
library(shiny)
library(twitteR)
library(plyr)
library(httr)
library(RJSONIO)
source("../twitter_stuff.R")
source("../Watson_NLU.R")
# Define UI for application that draws a histogram
ui <- fluidPage(
title = "Tweet Analysis with Watson",
sidebarLayout(
sidebarPanel(
conditionalPanel('input.dataset === "df"',
checkboxGroupInput("show_vars",
"Tweet DB columns",
names(df),
selected = names(df)))),
mainPanel(
textInput("text",
label = h3("What do you want to search about?"),
value = " "),
actionButton(inputId = "action",label = "Go"),
hr(),
fluidRow(column(3, verbatimTextOutput("value"))),
tabsetPanel(id = 'dataset',
tabPanel("table_summary", DT::dataTableOutput("table_summary")),
tabPanel("df",DT::dataTableOutput("table_tweets"))
)
)
)
)
# Define server logic required to draw a histogram
server <- function(input, output) {
twitt_df <- twitter_stuff(1000)
twFeed_csv<- write.csv(twitt_df,file = "savedTweets.csv",append = TRUE)
df = twitt_df[sample(nrow(twitt_df),500),]
bind_Signal <- data.frame() #take tweet id from tweet dataframe and bind POST content with the tweet_id
bind_Signal_to_df<-data.frame()
#choose columns to display
jsonSignal<-give_me_a_json_damnit(df, bind_Signal, bind_Signal_to_df)
output$table_summary<- DT::renderDataTable({DT::datatable(jsonSignal)})
output$table_tweets<- DT::renderDataTable({DT::datatable(df,options = list(lengthMenu = c(50, 300, 500)))})
output$value <- renderPrint("Done")
}
# Run the application
shinyApp(ui = ui, server = server)
#add twitter keyword as input parameter from shiny textbox at later stage
library(twitteR)
library(plyr)
twitter_stuff<- function(searchTerm, numTweets){
#Setup the Twitter Account and pass the tokens
setup_twitter_oauth(APIKey,APISecret,accessToken,accessTokenSecret)
#input text "Dosa OR Dhosa OR #Dosa OR #Dhosa
tweets_list <- twitteR::searchTwitteR(searchString = searchTerm, n=numTweets,lang = "en")
#getUser<- twitteR::getUser("@Deals4Every")
tweets_text = laply(tweets_list, function(t) t$getText())
tweets_clean <- clean.text(tweets_text)
#convert the tweets_text object to a data frame
#make data frame
df <- do.call("rbind", lapply(tweets_list, as.data.frame))
df["text"] = tweets_clean
return(df)
}
View(mat)
View(mat.df)
mat[1]
mat[1][1]
mat[:1]
mat[,:1]
mat[,1:]
mat[1:]
mat$entities
mat$entities[1]
mat$entities[1].unlist()
mat$entities[1]
mat$entities[[1]
]
mat$entities[7]
unlist(mat$entities[7])
unlist(mat$entities[8])
x <-unlist(mat$entities[8])
as.data.frame(x)
y<- data.frame(x)
View(y)
y<- do.call("rbind",lappy(x, as.data.frame)
)
y<- do.call("rbind",lapply(x, as.data.frame))
y<- do.call("cbind",lapply(x, as.data.frame))
y<- data.frame(x)
transform(y)
y<-transform(y)
library(reshape2)
foo<- cbind(ID=rownames(foo),foo)
foo<- cbind(ID=rownames(y),y)
melt(foo)
y<- melt(foo)
View(y)
y<-data.frame()
foo<-data.frame()
x <-unlist(mat$entities[8])
y<- as.data.frame(t(x))
clear
View(y)
View(mat)
View(mat.df)
foo<-lapply(mat$keywords,FUN = function(x) unlist(x))
tarzan<- as.data.frame(t(foo))
View(tarzan)
tarzan<- as.data.frame(foo)
foo<-lapply(mat$keywords,FUN = function(x) unlist(x))
tarzan<- as.data.frame(foo)
foo<-lapply(mat$keywords,FUN = function(x) unlist(x))
t(tarzan)
t(tarzan)
View(tarzan)
foo[1]
foo[2]
foo[3]
foo[4]
mat$keywords[[1]]
mat$keywords[[8]]
mat$categories[[8]]
mat$keywords[[8]]$text
mat$keywords[[8]]$text[1]
mat$keywords[[8]]
id_supply<- function(){}
id_supply<- function(){}
id_supply<- function(){  }
mat$keywords[[8]]$text[1]
install.packages(c("hash", "hashFunction", "hashids", "hashmap"))
h<- hash::hash()
hash::values(h,c=(mat$keywords[[8]]$text[1], mat$keywords[[8]]$text[2]))
hash::values(h,c=(mat$keywords[[8]]$text[1]))
mr<- hash::values(h,c=(mat$keywords[[8]]$text[1]))
mr
h <- hash( keys=letters, values=1:26 )
library(hash)
h <- hash( keys=letters, values=1:26 )
h
mat$keywords[[8]]$text[1]
id_supply <- function(df){
arr <- df$keywords
set.seed(42)
word_vector <- vector()
for(i in 1:length(arr)){
word_vector<- df$keywords[[i]]
}
return(word_vector)
}
id_supply(mat)
id_supply(mat)
debugSource('F:/R_Projects/Twitter_Trends/word_matrix.R')
View(word_vector)
View(word_vector)
id_supply <- function(df){
arr <- df$keywords
set.seed(42)
word_vector <- vector()
temp_vec<- vector()
temp_vec2<- vector()
for(i in 1:length(arr)){
word_vector<- df$keywords[[i]]
temp_vec <- word_vector
}
temp_vec2 <- cbind(temp_vec,temp_vec2)

Tweet_Data/app.R
library(shiny)
library(twitteR) 
library(plyr)
library(httr)
library(RJSONIO)
source("../twitter_stuff.R")
source("../Watson_NLU.R")


# Define UI for application that draws a histogram
ui <- fluidPage(
    title = "Tweet Analysis with Watson",
    sidebarLayout(
          sidebarPanel(
              conditionalPanel('input.dataset === "df"',
                  checkboxGroupInput("show_vars", 
                                     "Tweet selectors",
                                      names(df), 
                                      selected = names(df)))),
    
    mainPanel(
      textInput("text", 
                label = h3("What do you want to search about?"), 
                value = " "),
      actionButton(inputId = "action",label = "Go"),
      hr(),
      fluidRow(column(3, verbatimTextOutput("value"))),
      tabsetPanel(id = 'dataset',
                  tabPanel("table_summary", DT::dataTableOutput("table_summary")),
                  tabPanel("df",DT::dataTableOutput("table_tweets"))
                 ))))


searchTerm <- "#Dosa OR Dosa"
twitt_df<- data.frame()


# Define server logic required to draw a histogram
server <- function(input, output) {
  twitt_df <- twitter_stuff(searchTerm,500)
  twFeed_csv<- write.csv(twitt_df,file = "savedTweets.csv",append = TRUE)
  df = twitt_df[sample(nrow(twitt_df),500),]
  #choose columns to display
    jsonSignal<-NLU.results(df)
    write.csv(x = jsonSignal, file= "jsonSignal_TweetID.csv",append = TRUE)
    output$table_summary<- DT::renderDataTable({DT::datatable(jsonSignal)})
    output$table_tweets<- DT::renderDataTable({DT::datatable(df,
                          options = list(lengthMenu = c(50, 300, 500)))})
     output$value <- renderPrint("Done") 
    }
# Run the application 
shinyApp(ui = ui, server = server)

